sway/input/seatop_touch.c

```diff
@@ -0,0 +1,110 @@
+#define _POSIX_C_SOURCE 200809L
+#include <wlr/types/wlr_touch.h>
+#include <wlr/types/wlr_cursor.h>
+#include "sway/input/seat.h"
+#include "sway/input/cursor.h"
+#include "log.h"
+#include "config.h"
+
+struct seatop_touch_point_event {
+	double ref_lx, ref_ly;         // cursor's x/y at start of op
+	double ref_con_lx, ref_con_ly; // container's x/y at start of op
+	int32_t touch_id;
+	struct wl_list link;
+};
+
+struct seatop_touch_event {
+	//struct sway_container *con;
+	struct wl_list point_events;
+};
+
+static void handle_touch_down(struct sway_seat *seat,
+		struct wlr_event_touch_down *event) {
+	fprintf(stderr, "handle_touch_down !\n");
+
+	struct sway_cursor *cursor = seat->cursor;
+	struct wlr_surface *surface = NULL;
+	struct seatop_touch_event *e = seat->seatop_data;
+
+	double lx, ly;
+	wlr_cursor_absolute_to_layout_coords(cursor->cursor, event->device,
+			event->x, event->y, &lx, &ly);
+	double sx, sy;
+	node_at_coords(seat, lx, ly, &surface, &sx, &sy);
+
+	if (!surface) { // Must start from the surface
+		return;
+	}
+
+	struct seatop_touch_point_event *point_event = calloc(1, sizeof(struct seatop_touch_point_event));
+	point_event->ref_lx = lx;
+	point_event->ref_ly = ly;
+	point_event->ref_con_lx = sx;
+	point_event->ref_con_ly = sy;
+	point_event->touch_id = event->touch_id;
+	wl_list_insert(&e->point_events, &point_event->link);
+}
+
+static void handle_touch_up(struct sway_seat *seat,
+		struct wlr_event_touch_up *event) {
+	fprintf(stderr, "handle_touch_up !\n");
+	struct seatop_touch_event *e = seat->seatop_data;
+
+	struct seatop_touch_point_event *point_event;
+	wl_list_for_each(point_event, &e->point_events, link) {
+		if (point_event->touch_id == event->touch_id) {
+			fprintf(stderr, "We found our point event\n");
+			break;
+		}
+	};
+
+	wl_list_remove(&point_event->link);
+
+	if (wl_list_empty(&e->point_events)) {
+		//seatop_end(seat);
```

Comment:
This make sway to crash. I'm not sure yet how to release cleanly the seatop.

Reply:
To change back to the default seatop, use `seatop_begin_default`. See `handle_button` in `sway/input/seatop_down.c` for an example.

Reply:
Yeah, way better ! thanks !